rfcs/geometric-primitives.md

```diff
@@ -0,0 +1,349 @@
+  extent: Vec3,
+}
+
+struct Cone{
+  /// Origin of the cone located at the center of its base.
+  origin: Point,
+  /// The extent of the cone is the vector that extends from the origin to the tip of the cone.
+  extent: Vec3,
+  base_radius: f32,
+}
+
+struct Torus{
+  origin: Point,
+  normal: Direction,
+  major_radius: f32,
+  tube_radius: f32,
+}
+
+struct Capsule {
+  origin: Point,
+  height: Vec3,
+
+}
+
+// A 3d frustum used to represent the volume rendered by a camera, defined by the 6 planes that set the frustum limits.
+struct Frustum{
+  near: Plane,
+  far: Plane,
+  top: Plane,
+  bottom: Plane,
+  left: Plane,
+  right: Plane,
+}
+
+// 3D Axis Aligned Bounding Box defined by its extents, useful for fast intersection checks and frustum culling.
+struct AabbExtents {
+  min: Vec3
+  max: Vec3
+}
+impl Aabb for AabbExtents {} //...
+
+// 3D Axis Aligned Bounding Box defined by its center and half extents, easily converted into an OBB
+struct AabbCentered {
+  origin: Point,
+  half_extents: Vec3,
+}
+impl Aabb for AabbCentered {} //...
+
+// 3D Axis Aligned Bounding Box defined by its eight vertices, useful for culling or drawing
+struct AabbVertices([Point; 8]);
+impl Aabb for AabbVertices {} //...
+
+// 3D Oriented Bounding Box
+struct ObbCentered {
+  origin: Point,
+  orthonormal_basis: Mat3,
+  half_extents: Vec3,
+}
+impl Obb for ObbCentered {} //...
+
+struct ObbVertices([Point; 4]);
+impl Obb for ObbVertices {} //...
+```
+
+### 2D Geometry Types
+
+These types only exist in 2d space: their dimensions and location are only defined in `x` and `y` unlike their 3d counterparts. These types are suffixed with "2d" to disambiguate from the 3d types in user code, guide users to using 3d types by default, and remove the need for name-spacing the 2d and 3d types when used in the same scope.
+
+```rust
+/// Point in 2D space
+struct Point2d(Vec2)
+
+/// Vector direction in 2D space that is guaranteed to be normalized through its getter/setter.
+struct Direction2d(Vec2)
+
+/// Unbounded line in 2D space with direction
+struct Ray2d { 
+  point: Point2d, 
+  normal: Direction2d,
+}
+
+/// Line in 2D space bounded by two points
+struct Line2d { 
+  start: Point2d, 
+  end: Point2d,
+}
+
+/// A line list represented by an ordered list of vertices in 2d space
+struct LineList<const N: usize>{
+  points: [Point; N],
+  /// True if the LineList is a closed loop
+  closed: bool,
+}
+
+struct Triangle2d([Point2d; 3]);
+
+struct Quad2d([Point2d; 4]);
+
+/// A regular polygon, such as a square or hexagon.
+struct RegularPolygon2d{
+  circumcircle: Circle2d,
+  /// Number of faces.
+  faces: u8,
+  /// Clockwise rotation of the polygon about the origin. At zero rotation, a point will always be located at the 12 o'clock position.
+  orientation: Angle,
+}
+
+struct Circle2d {
+  origin: Point2d, 
+  radius: f32,
+}
+
+/// Segment of a circle
+struct Arc2d {
```

Comment:
3d version of an arc seems to be missing

Reply:
Yeah, I think this warrants more discussion. I'm unsure if there is utility in defining 2d shapes in 3d space. I've been thinking about this, and will plan to write a description soon.

Reply:
Personally, I'd find a lot of value in that for various games, even if the engine itself doesn't have much immediate use. You could always combine arbitrary 2D shapes with a plane I suppose, but that's a frustrating representation to work with and think about.

Reply:
I ended up removing (see comment on ellipses). This monstrosity is already feeling pretty hefty, so I'd like to cut down on types that aren't immediately useful for constructing colliders.